### PR TITLE
[GlobalOpt][Flow] Add GlobalOp folding to FoldUnitExtentDims

### PR DESCRIPTION
**Pull request by Max191**
This adds a new pass to fold all unit dims on mutable globals. Reshape ops are inserted at the loads and stores, which should ideally fold with unit dim reshapes introduced in Flow.

### CODE REVIEW

**Review by qedawkins — COMMENTED** [review]
is there a reason to do this with a separate pass instead of just doing this at the same time as `LinalgFoldUnitExtentDims`? Is it a matter of this pass needing to nest on the module and the linalg version nests on functions?

**Comment by Max191** [comment]
> is there a reason to do this with a separate pass instead of just doing this at the same time as `LinalgFoldUnitExtentDims`? Is it a matter of this pass needing to nest on the module and the linalg version nests on functions?

Well the main reason I put it in GlobalOptimization is so it will happen before the analysis in https://github.com/openxla/iree/pull/16585

EDIT: Wait, I didn't realize `LinalgFoldUnitExtentDims` was moved to global optimization. I guess it could be part of that pass, but there is the different nesting as you mentioned.

**Comment by qedawkins** [comment]
> > is there a reason to do this with a separate pass instead of just doing this at the same time as `LinalgFoldUnitExtentDims`? Is it a matter of this pass needing to nest on the module and the linalg version nests on functions?
> 
> Well the main reason I put it in GlobalOptimization is so it will happen before the analysis in #16585
> 
> EDIT: Wait, I didn't realize `LinalgFoldUnitExtentDims` was moved to global optimization. I guess it could be part of that pass, but there is the different nesting as you mentioned.

Yeah I'll let others comment, but I think nesting is probably enough justification (there's nothing those patterns need from the GreedyPatternRewriter so should net a little extra parallelism?)

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: This should also skip public globals

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: Can you use the Explorer to set up this map? It might be overkill in this case, and you don't need the inverse call graph, but I'd guess that the code would look more canonical? You can use `Explorer::initializeGlobalInfos` on its own to initialize the global information without setting up the callgraph I think.

Also provides nice APIs like `forEachGlobal` and caches information like mutability, indirect accessors, and visibility.

Comment (Max191) [reply compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: Oh nice, I didn't know `Explorer::initializeGlobalInfos` already does this. I'll use that.

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: Maybe using `clone` (https://github.com/shark-infra/llvm-project/blob/3c25a9de91e9c76ffba82939b19eafe3e60d51f7/mlir/include/mlir/Dialect/Utils/StructuredOpsUtils.h#L142) will make this easier?

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: Again the `clone` method` above might help.

**Review by MaheshRavishankar — APPROVED** [review]
My sense is that this being done way too early. Agree with Quinn, maybe fold this into the `FoldUnitDims` pass.

I am stamping cause overall this looks fine, but please address all comments before landing.

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: Can you declare it in the beginning of the function, so the `IRRewriter` can also use it?

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: Please add a test for this case.

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: can this be `globalType.clone(newShape)`?

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: The init value is `std::nullopt`, so you dont have to specify it.

https://en.cppreference.com/w/cpp/utility/optional/optional

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: Let's do it in the other way, which could save level of indents.

```cpp
auto uninitializedAttr = dyn_cast<...>(...);

// TODO...
if (!uninitializedAttr) return success();

newInitialValue = ...
```

Also, is it possible that `getGlobalInitialValue()` is NULL? I think we don't need `or_null` here. And please drop `llvm::`.

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: +1

File: compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp]: I usually add `OpBuilder::InsertionGuard guard(rewriter);` right before setting insertion point. I think it is callee's responsibility to restore the state.

**Review by hanhanW — APPROVED** [review]

**Comment by Max191** [comment]
> My sense is that this being done way too early. Agree with Quinn, maybe fold this into the `FoldUnitDims` pass.

I decided to just change the nesting of FoldUnitExtentDims and add the GlobalOp folding there. It saves a barrier for reshape canonicalization pattern rewriting.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FoldUnitExtentDims.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FoldUnitExtentDims.cpp:110]: I think you can do the global info initialization on its own right?

**Review by qedawkins — APPROVED** [review]
nice, LGTM

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FoldUnitExtentDims.cpp
Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FoldUnitExtentDims.cpp:110]: No, that's a private function. I guess I could make it public, though. @benvanik do you care if I make `Explorer::initializeGlobalInfos` public so I don't have to initialize the inverseCallGraph?

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**
